core/io/image.cpp

```diff
@@ -3929,22 +3929,79 @@ void Image::renormalize_rgbe9995(uint32_t *p_rgb) {
 	// Never used
 }
 
+Image::FileFormat Image::detect_file_format(const uint8_t *p_data, int p_len) {
+	size_t len = p_len < 0 ? 0 : unsigned(p_len);
+
+	static const uint8_t png_file_header[] = { 0x89, 0x50, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A }; // \211PNG\r\n\032\n
+	static const size_t png_file_header_size = sizeof(png_file_header);
+	if ((!len || len > png_file_header_size) && memcmp(p_data, png_file_header, png_file_header_size) == 0) {
+		return FILE_FORMAT_PNG;
+	}
```

Comment:
I have a feeling this code should be in the respective modules for the image type. The current design of Image avoids having filetype-specific code by offloading all the work to the various modules. This is in line with our engine best practice to ["prefer local solutions"](https://docs.godotengine.org/en/latest/contributing/development/best_practices_for_engine_contributors.html#prefer-local-solutions). 

We are especially strict with core types (like Image), because if they become bloated and messy, then the entire engine becomes bloated and messy.

Reply:
This method is used in the image constructor, which takes only raw data as input.

This constructor is only used in one place - [texture.cpp PortableCompressedTexture2D::_set_data](https://github.com/godotengine/godot/blob/a83eb16fba5bb3da086b41cbd79e6f95b09eb8ee/scene/resources/texture.cpp#LL374C31-L374C31). I'd really like to move the image type detection function to a different location. Can you please tell me what is the best place for this?

Reply:
I would add it to the respective loader functions i.e. ``_png_mem_loader_func``. If the format is the wrong type those functions can exit early which will leave the image empty so that the next format can be checked.

Reply:
done